plugin/trino-kafka-event-listener/src/main/java/io/trino/plugin/eventlistener/kafka/KafkaEventPublisher.java

```diff
@@ -0,0 +1,149 @@
+        String completedTopic = config.getCompletedTopicName().orElse("");
+
+        Map<String, String> configOverrides = config.getKafkaClientOverrides();
+        LOG.info("Kafka Event Listener: Creating Kafka publisher (SSL=%s) for topics: %s/%s with excluded fields: %s and kafka config overrides: %s",
+                producerFactory instanceof SSLKafkaProducerFactory, startedTopic, completedTopic, config.getExcludedFields(), configOverrides);
+        kafkaProducer = producerFactory.producer(configOverrides);
+        checkConnectivityToBrokers(config.getPublishCreatedEvent() ? startedTopic : completedTopic, config.getRequestTimeout());
+        kafkaRecordBuilder = new KafkaRecordBuilder(startedTopic, completedTopic, config.getExcludedFields(), metadataProvider(config));
+        LOG.info("Kafka Event Listener: Successfully created Kafka publisher.");
+    }
+
+    private void checkConnectivityToBrokers(String topic, int requestTimeout)
+            throws Exception
+    {
+        try {
+            LOG.debug("Kafka Event Listener: checking connectivity to brokers (fetching partitions for topic=%s).", topic);
+            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> kafkaProducer.partitionsFor(topic));
+            future.get(requestTimeout, TimeUnit.MILLISECONDS);
+            LOG.debug("Kafka Event Listener: connectivity check succeeded.");
+        }
+        catch (Exception e) {
+            LOG.error("Kafka Event Listener: connectivity check failed: %s", e.getMessage());
+            throw e;
+        }
+    }
+
+    private MetadataProvider metadataProvider(KafkaEventListenerConfig config)
+    {
+        if (config.getEnvironmentVariablePrefix().isPresent()) {
+            return new EnvMetadataProvider(config.getEnvironmentVariablePrefix().get());
+        }
+        else {
+            return new NoOpMetadataProvider();
+        }
+    }
+
+    public void publishCompletedEvent(QueryCompletedEvent queryCompletedEvent)
+    {
+        stats.completedEventReceived();
+        String queryId = queryCompletedEvent.getMetadata().getQueryId();
+        LOG.debug("Kafka Event Listener: preparing to send QueryCompletedEvent for query id: %s", queryId);
+        ProducerRecord<String, GenericRecord> record = null;
+        try {
+            record = kafkaRecordBuilder.buildCompletedRecord(queryCompletedEvent);
+        }
+        catch (Exception e) {
+            stats.completedEventBuildFailure();
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(e, "Kafka Event Listener: unable to build QueryCompletedEvent for query id: %s", queryId);
+            }
+            else {
+                LOG.warn("Kafka Event Listener: unable to build QueryCompletedEvent for query id: %s. Error msg: %s", queryId, e.getMessage());
+            }
+        }
+        if (record != null) {
+            kafkaProducer.send(record, (metadata, exception) -> {
```

Comment:
consider that the event listeners block the engine from making progress until they return.

There were ideas in past to make event listener authors choose if they want async processing or blocking sync processing. Do you think it's a good idea to revisit this? Since you're pushing to Kafka I assume being async/batched doesn't matter for this case?

Reply:
> consider that the event listeners block the engine from making progress until they return.

I think it's both.
IIRC `queryCreated` will be blocking
while `splitCompleted` is using unbounded executor.
Both are not perfect. To be double checked

Reply:
Create PR to cap concurrent completed requests: https://github.com/trinodb/trino/pull/22911

Reply:
Is the queue in `KafkaProducer` bounded?

Reply:
Yes, looks like it's bounded: `KafkaProducer` [appends](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1071) the new event object to a `RecordAccumulator`.
`RecordAccumulator` [stores](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L350-L351) the object in a `BufferPool` internally, which is [bounded](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java#L75) by the `totalMemory` parameter.